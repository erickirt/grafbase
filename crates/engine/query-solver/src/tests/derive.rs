use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
    type Product
      @join__type(graph: EXT)
    {
      authorId: ID!
      commentIds: [ID!]!
      inventoryCountryId: ID!
      inventoryWarehouseId: ID!
      inventoriesKeys: [InventoryKeys!]!
      code: String!
      id: ID!
      author: User! @composite__derive(graph: EXT)
      comments: [Comment!]! @composite__derive(graph: EXT)
      inventory: Inventory! @composite__derive(graph: EXT)
      inventories: [Inventory!]! @composite__derive(graph: EXT)
    }

    type User
        @join__type(graph: EXT, key: "id", resolvable: false)
    {
      id: ID!
    }

    type Comment
        @join__type(graph: EXT, key: "id", resolvable: false)
    {
      id: ID!
    }

    type InventoryKeys @join__type(graph: EXT) {
        countryId: ID!
        warehouseId: ID!
    }

    type Inventory
        @join__type(graph: EXT, key: "countryId warehouseId", resolvable: false)
    {
        countryId: ID!
        warehouseId: ID!
    }

    type Query
    {
      products: [Product!]! @join__field(graph: EXT)
    }

    enum join__Graph
    {
    EXT @join__graph(name: "ext", url: "http://localhost:8080")
    }
"#;

#[test]
fn single_id() {
    assert_solving_snapshots!(
        "single_id",
        SCHEMA,
        r#"
        query {
            products {
                author {
                    id
                }
            }
        }
        "#
    );
}

#[test]
fn composite_keys() {
    assert_solving_snapshots!(
        "composite_keys",
        SCHEMA,
        r#"
        query {
            products {
                 inventory {
                    countryId
                    warehouseId
                }
            }
        }
        "#
    );
}

#[test]
fn batch_single_id() {
    assert_solving_snapshots!(
        "batch_single_id",
        SCHEMA,
        r#"
        query {
            products {
                comments {
                    id
                }
            }
        }
        "#
    );
}

#[test]
fn batch_composite_keys() {
    assert_solving_snapshots!(
        "batch_composite_keys",
        SCHEMA,
        r#"
        query {
            products {
                inventories {
                    countryId
                    warehouseId
                }
            }
        }
        "#
    );
}
