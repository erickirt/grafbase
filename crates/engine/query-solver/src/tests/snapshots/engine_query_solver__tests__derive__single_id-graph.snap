---
source: crates/engine/query-solver/src/tests/derive.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Query.products\" ]\n    2 [ label = \"Product.author\" ]\n    3 [ label = \"User.id\" ]\n    4 [ label = \"Root#ext\", shape=parallelogram, color=dodgerblue ]\n    5 [ label = \"products#ext\", shape=box, color=dodgerblue ]\n    6 [ label = \"author#ext@derive\", shape=box, color=dodgerblue ]\n    7 [ label = \"id#ext@derive\", shape=box, color=dodgerblue ]\n    0 -> 1 [  ]\n    1 -> 2 [  ]\n    2 -> 3 [  ]\n    0 -> 4 [ color=royalblue,fontcolor=royalblue ]\n    0 -> 4 [ style=dashed,arrowhead=none ]\n    4 -> 5 [ color=royalblue,fontcolor=royalblue ]\n    5 -> 1 [ color=violet,arrowhead=none ]\n    5 -> 6 [ color=royalblue,fontcolor=royalblue ]\n    6 -> 2 [ color=violet,arrowhead=none ]\n    6 -> 7 [ color=royalblue,fontcolor=royalblue ]\n    7 -> 3 [ color=violet,arrowhead=none ]\n}\n"
---
digraph {
    0 [ root]
    1 [ Query.products]
    2 [ Product.author]
    3 [ User.id]
    4 [ Root#ext]
    5 [ products#ext]
    6 [ author#ext@derive]
    7 [ id#ext@derive]
    0 -> 1 [ label = "Field" ]
    1 -> 2 [ label = "Field" ]
    2 -> 3 [ label = "Field" ]
    0 -> 4 [ label = "CreateChildResolver" ]
    0 -> 4 [ label = "HasChildResolver" ]
    4 -> 5 [ label = "CanProvide" ]
    5 -> 1 [ label = "Provides" ]
    5 -> 6 [ label = "CanProvide" ]
    6 -> 2 [ label = "Provides" ]
    6 -> 7 [ label = "CanProvide" ]
    7 -> 3 [ label = "Provides" ]
}
